File: src/structures/Point2d.java
package structures;

/**
 * @author Sean Lewis
 */
public class Point2d {

	/**
	 * The origin point (0,0).
	 */
	public static final Point2d ORIGIN = new Point2d(0.0, 0.0);

	/**
	 * The x-coordinate of this point.
	 */
	public double x;

	/**
	 * The y-coordinate of this point.
	 */
	public double y;

	/**
	 * Creates a new point at (0.0).
	 */
	public Point2d() {
		x = 0.0;
		y = 0.0;
	}

	/**
	 * Creates the point at (x,y)
	 * @param x the x-coordinate
	 * @param y the y-coordinate
	 */
	public Point2d(double x, double y) {
		this.x = x;
		this.y = y;
	}

	/**
	 * Creates a new point based on p's coordinates
	 * @param p a java.awt.Point
	 */
	public Point2d(java.awt.Point p) {
		this.x = p.x;
		this.y = p.y;
	}

	/**
	 * Creates a integer-based point by rounding this point's coordinates.
	 * @return a point with rounded coordinates
	 */
	public java.awt.Point getIntegerPoint() {
		return new java.awt.Point((int) Math.round(x), (int) Math.round(y));
	}

	/**
	 * Returns the he distance between this and another point.
	 * @param p another point
	 * @return the the distance between points.
	 */
	public double getDistance(Point2d p) {
		return Math.hypot(x - p.x, y - p.y);
	}

	/**
	 * Returns the square of the distance between this and another point.
	 * @param p another point
	 * @return the square of the distance between points.
	 */
	public double getDistanceSquared(Point2d p) {
		return Math.pow(x - p.y, 2.0) + Math.pow(y - p.y, 2.0);
	}

	/**
	 * Returns the translation of this point by dx and dy.
	 * @param dx the x-difference
	 * @param dy the y-difference
	 * @return the translated point
	 */
	public Point2d translate(double dx, double dy) {
		return new Point2d(x + dx, y + dy);
	}

	/**
	 * Returns the x component
	 * @return the x component
	 */
	public double getX() {
		return x;
	}

	/**
	 * Sets the x component
	 * @param x a parameter
	 */
	public void setX(double x) {
		this.x = x;
	}

	/**
	 * Returns the y component
	 * @return the y component
	 */
	public double getY() {
		return y;
	}

	/**
	 * Sets the y component
	 * @param y a parameter
	 */
	public void setY(double y) {
		this.y = y;
	}

	/**
	 * Returns a mathematical, String representation of this point in the form
	 * (x, y).
	 */
	public String toString() {
		return "(" + x + ", " + y + ")";
	}

	/**
	 * Returns if two points have the same perpendicular components, to within
	 * epsilon precision.
	 * @param p a point
	 * @param epsilon maximum allowed difference
	 * @return if the objects represent the same point
	 */
	public boolean equals(Point2d p, double epsilon) {
		double xDif = x - p.x;
		double yDif = y - p.y;
		return Math.abs(xDif) <= epsilon && Math.abs(yDif) <= epsilon;
	}

}

File: src/structures/Point2d.scala
package structures

/** An immutable double-precision 2-d Cartesian point.
 * 
 * @constructor create a new point at the given coordinates
 * @param x the x coordinate
 * @param y the y coordinate
 */
class Point2d(val x: Double, val y: Double) {

  /** Creates a point at the origin. */
  def this() = this(0.0, 0.0)

  /** Creates a point given by the input point. */
  def this(p: java.awt.Point) = this(p.x, p.y)
  
  /** Creates a point as determined by a vector. */
  def this(v: Vector2d) = this(v.xComponent, v.yComponent)

  /** Rounds the point to an integer-precision point. */
  def getIntegerPoint(): java.awt.Point = {
    def round(x: Double): Int = math.round(x.asInstanceOf[Float])
    new java.awt.Point(round(x), round(y))
  }
  
  /** Gives the square of the distance between two points. */
  def distanceSquared(p: Point2d): Double = {
    val dx = x - p.x
    val dy = y - p.y
    dx * dx + dy * dy
  }

  /** Gives the distance between two points. */
  def distance(p: Point2d): Double = math.sqrt(distanceSquared(p))

  /** Tells if this point is within a certain maximum distance from another point.*/
  def withinDistance(p: Point2d, max: Double): Boolean =
    distanceSquared(p) <= max * max

  /** Translates a point by dx and dy. */
  def translate(dx: Double, dy: Double): Point2d = new Point2d(x + dx, y + dy)

  /** Translates a point by dx and dy. */
  def translate(p: Point2d): Point2d = translate(p.x, p.y)

  /** Translates a point by dx and dy. */
  def +(dx: Double, dy: Double): Point2d = translate(dx, dy)

  /** Translates a point by the values of another point */
  def +(p: Point2d): Point2d = translate(p.x, p.y)

  /** Translates a point by -dx and -dy. */
  def -(dx: Double, dy: Double): Point2d = translate(-dx, -dy)

  /** Translates a point by the negation of another point */
  def -(p: Point2d): Point2d = this - (p.x, p.y)

  /** Tells if another instance can equal a Point2d. */
  def canEqual(other: Any): Boolean = other.isInstanceOf[structures.Point2d]  

  override def equals(other: Any): Boolean = {
    other match {
      case that: structures.Point2d => 
        that.canEqual(Point2d.this) &&
        x == that.x && y == that.y
      case _ => false
    }
  }
  
  /** Tells if two types are equal, with epsilon error allowed. */
  def equals(other: Any, epsilon: Double): Boolean = {
    def eq(x: Double, y: Double) = math.abs(x - y) <= epsilon
    other match {
      case that: structures.Point2d => 
        that.canEqual(Point2d.this) &&
        eq(x, that.x) &&
        eq(y, that.y)
      case _ => false
    }
  }
  
  override def hashCode(): Int = {
    val prime = 41
    prime * (prime + x.hashCode) + y.hashCode
  }

  override def toString(): String = "(" + x + ", " + y + ")";

}